fix: improve portability of source tree

aronson opened the pull request:
Hi, I tried to run this on mobile using Android and Termux, as well as run this on macOS with my Mac and Apple clang, and it didn't build.

I updated a few things to do this:
- I changed the submodule references to use HTTPS instead of SSH so that inexperienced users can build without being logged in to GitHub
- I changed `pthread` invocations to support macOS when `__APPLE__` is defined by the compiler
- I removed the non-portable `argp.h` import and instead leveraged [a header-only library](https://github.com/Dugy/quick_arg_parser) to replace it. That `argp.h` header isn't available on macOS nor Android.
  - This library is a C++ header so I moved to C++11 which has decently wide support in 2025
- I modified the picoquic submodule:
  - Termux doesn't define `ANDROID` when used in this config, so I instead relied on the appropriate built-in definition per the POSIX specification
  - I removed the `-fno-exception` flags from picoquic's CMake definitions as it broke the build given quick_arg_parser leverages C++ exceptions
- I updated the `.gitignore` to ignore some files that were bugging me when working on this

EDIT: I also tested it, of course.

Server confirmed working on:
- Ubuntu 24.04 Linux x86_64

Client confirmed working (including multipath) on:
- Ubuntu 24.04 Linux x86_64
- Fedora latest Linux x86_64
- macOS 26 ARM64
- Android 16 Termux

EndPositive commented:
Hey @aronsonm thanks a lot for your PR! This cmake stuff is all a bit magic to me, so I'm happy to take your contributions. Can you verify that your latest version works on Mac in Docker? The Dockerfile doesn't seem to build anymore due to the `__ANDROID__` change. Are you sure Termux doesn't define `__ANDROID__`? It seems a popular fix for running things there.

If I revert that, the Dockerfile builds but doesn't run anymore, due to the final binary not missing `libstdc++.so.6`. I'd be really happy if we could statically build the final slipstream binaries in Docker.

aronson commented:
My findings were Docker was broken out of the box in the `bookworm-slim` builder set up for the `slipstream-picoquic` library because `make check` was leveraging test code TUs that had symbol/ABI drift. The correct fix concerning *my* code was to add `#include <signal.h>` to picoquic so that we can get the `SIGKILL` symbol. I bet the more modern compilers I was using add this to the default include path for us...
After that, and in the original `master` branch, there were missing initializer arguments for `param1` and `param2` in the test code. I set those to `0`. There was also another method that had some drift... I fixed those and ran the docker build...
With that change `make check` works now in the subproject and the `docker build` completes!

I think part of the reason the original `slipstream` didn't work in Termux is because `ANDROID` is not `__ANDROID__`. While I support the change in theory, it's still not correct as Android is not the only potential platform where `_POSIX_THREAD_PROCESS_SHARED` may or may not be defined (for example, tight embedded contexts, or maybe something using picolibc or nanolib with weird thread-local storage setups, I bet it's broken on PS Vita too... but it's how the spec defines it all...). I'd leave my change in and express the desired abstraction to use in a platform-independent way.

As for the missing `libstdc++.so.6`, I've added the right flags to link it in statically since that seems to be what you're looking for. The containers work once again:
```
❯ docker run --rm slipstream:latest --help
slipstream-server - A high-performance covert channel over DNS (server)

Usage: /usr/src/app/slipstream-server [options]
-l	--dns-listen-port	 DNS listen port (default: 53)
-a	--target-address	 Target server address (default: 127.0.0.1:5201)
-c	--cert	 Certificate file path (default: certs/cert.pem)
-k	--key	 Private key file path (default: certs/key.pem)
-d	--domain	 Domain name this server is authoritative for (Required)
```

Let me know how it looks now 🙂 

P.S.: I considered porting to meson for my own reasons, but I figured it would be kinda rude to change build systems under you with my first PR. If you're open to that, it's much easier to use than CMake in my experience. I re-tooled [Apotris](https://apotris.com) ages ago away from make to meson and it gave us bonkers level of portability.

aronson commented:
OK that fixed it all. I guess you'll need to only use a static link to the c++ library in Docker. Tested and working as expected in Docker, Termux, bare-metal Linux, and macOS.

aronson commented:
You can include CMake projects (though it's a little poorly documented). I leverage this in that other project to great success for an unreleased feature branch. I'll take a stab at it soon and open a PR leveraging a new release flow for you when I can! Stay tuned...
